File: waltid-libraries/protocols/waltid-wallet-sdk-ios/Sources/WalletSDKEnterpriseCustody/EnterpriseIdentityKeyCustodian.swift
import Foundation
import WalletSDK
#if os(iOS)

/// Optional Enterprise KMS import integration. It gives the KMS an additional private-key copy;
/// it does not preserve a recovery record, configure remote signing, or synchronize credentials.
public actor EnterpriseIdentityKeyCustodian: IdentityKeyCustodian {
    /// Stable integration identifier.
    public nonisolated let id: String
    /// Destination label shown before explicit selection.
    public nonisolated let displayName: String
    private let resourceURL: URL
    private let configuration: URLSessionConfiguration
    private let authorize: @Sendable (URLRequest) async throws -> URLRequest

    /// Configures an explicit HTTPS KMS destination. The authorizer receives a body-free request.
    /// - Parameters:
    ///   - kmsResourceURL: Resource URL such as `https://enterprise.example/v1/org.kms`.
    ///   - id: Stable integration identifier.
    ///   - displayName: Destination label shown before selection.
    ///   - configuration: Host TLS and timeout settings; redirects are always refused.
    ///   - authorize: Adds authentication without changing the destination URL or method.
    public init(kmsResourceURL: URL, id: String = "enterprise-kms", displayName: String = "Enterprise KMS",
                configuration: URLSessionConfiguration = .ephemeral,
                authorize: @escaping @Sendable (URLRequest) async throws -> URLRequest) {
        precondition(kmsResourceURL.scheme == "https" && kmsResourceURL.user == nil && kmsResourceURL.password == nil)
        precondition(kmsResourceURL.query == nil && kmsResourceURL.fragment == nil && !kmsResourceURL.path.isEmpty && !kmsResourceURL.path.hasSuffix("/"))
        precondition(!id.isEmpty && !displayName.isEmpty)
        self.resourceURL = kmsResourceURL
        self.id = id
        self.displayName = displayName
        self.configuration = configuration
        self.authorize = authorize
    }

    /// Imports idempotently under the identity's stable key ID. The SDK verifies the returned public key.
    /// - Parameters:
    ///   - identity: Original identity and stable key ID.
    ///   - privateJWK: Secret P-256 JWK; never logged or returned in the receipt.
    public func importKey(identity: SigningIdentity, privateJWK: Data) async throws -> IdentityCustodyReceipt {
        guard identity.keyID.range(of: "^[A-Za-z0-9_-]{1,256}$", options: .regularExpression) != nil,
              let resource = URL(string: "\(resourceURL.absoluteString).\(identity.keyID)") else {
            throw IdentityProviderError.rejected
        }
        let endpoint = resource.appendingPathComponent("kms-service-api/keys/import/jwk")
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request = try await authorize(request)
        guard request.url == endpoint && request.httpMethod == "POST" else { throw IdentityProviderError.rejected }
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = privateJWK
        let session = URLSession(configuration: configuration, delegate: RefuseRedirects(), delegateQueue: nil)
        defer { session.invalidateAndCancel() }
        do {
            let (stream, response) = try await session.bytes(for: request)
            guard let http = response as? HTTPURLResponse else { throw IdentityProviderError.rejected }
            switch http.statusCode {
            case 200, 201: break
            case 401: throw IdentityProviderError.interactionRequired
            case 403: throw IdentityProviderError.rejected
            case 409: throw IdentityProviderError.conflict
            case 429, 500...599: throw IdentityProviderError.temporarilyUnavailable
            default: throw IdentityProviderError.rejected
            }
            var body = Data()
            for try await byte in stream {
                guard body.count < 65_536 else { throw IdentityProviderError.rejected }
                body.append(byte)
            }
            defer { body.resetBytes(in: 0..<body.count) }
            guard let view = (try? JSONSerialization.jsonObject(with: body)) as? [String: Any],
                  let key = view["key"] as? [String: Any], let jwk = key["jwk"] as? [String: Any] else {
                throw IdentityProviderError.rejected
            }
            let members = ["kty", "crv", "x", "y"]
            var publicKey: [String: String] = [:]
            for member in members {
                guard let value = jwk[member] as? String else { throw IdentityProviderError.rejected }
                publicKey[member] = value
            }
            guard publicKey["kty"] == "EC", publicKey["crv"] == "P-256",
                  let expected = (try? JSONSerialization.jsonObject(with: privateJWK)) as? [String: Any],
                  expected["kty"] as? String == "EC", expected["crv"] as? String == "P-256" else {
                throw IdentityProviderError.rejected
            }
            for coordinate in ["x", "y"] {
                guard let actual = publicKey[coordinate].flatMap(Self.coordinate),
                      let original = (expected[coordinate] as? String).flatMap(Self.coordinate) else {
                    throw IdentityProviderError.rejected
                }
                guard actual == original else { throw IdentityProviderError.conflict }
            }
            let publicData = try JSONSerialization.data(withJSONObject: publicKey, options: [.sortedKeys])
            return .init(keyReference: resource.absoluteString, publicJWK: String(decoding: publicData, as: UTF8.self))
        } catch is CancellationError { throw CancellationError() }
        catch let error as IdentityProviderError { throw error }
        catch {
            try Task.checkCancellation()
            throw IdentityProviderError.temporarilyUnavailable
        }
    }
    private static func coordinate(_ value: String) -> Data? {
        let base64 = value.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        guard let bytes = Data(base64Encoded: base64 + String(repeating: "=", count: (4 - base64.count % 4) % 4)),
              bytes.count == 32 else { return nil }
        return bytes
    }
}

private final class RefuseRedirects: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse,
                    newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
        completionHandler(nil)
    }
}
#endif

File: waltid-libraries/protocols/waltid-openid4vc-wallet-custody-enterprise/src/commonMain/kotlin/id/walt/wallet2/custody/enterprise/EnterpriseIdentityKeyCustodian.kt
package id.walt.wallet2.custody.enterprise

import id.walt.crypto2.keys.EcCurve
import id.walt.crypto2.keys.EncodedKey
import id.walt.crypto2.keys.KeySpec
import id.walt.crypto2.keys.toSpkiDer
import id.walt.crypto2.serialization.BinaryData
import id.walt.wallet2.mobile.identity.*
import io.ktor.client.HttpClient
import io.ktor.client.request.preparePost
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsChannel
import io.ktor.http.*
import io.ktor.http.content.TextContent
import io.ktor.utils.io.readRemaining
import kotlinx.coroutines.CancellationException
import kotlinx.io.readByteArray
import kotlinx.serialization.json.*

/** Optional Enterprise KMS import integration. It gives the KMS private-key custody; it does not
 * preserve a portable recovery record, synchronize credentials, or configure remote signing.
 *
 * Supply an authenticated client without request/response body logging. This adapter disables
 * redirects and closes only its configured client when [close] is called.
 * @param httpClient Host-owned authentication, TLS and timeout configuration.
 * @param kmsResourceUrl HTTPS KMS resource URL, e.g. `https://enterprise.example/v1/org.kms`.
 * @property id Stable integration identifier.
 * @property displayName Destination label presented before explicit selection. */
public class EnterpriseIdentityKeyCustodian(
    httpClient: HttpClient,
    private val kmsResourceUrl: Url,
    override val id: String = "enterprise-kms",
    override val displayName: String = "Enterprise KMS",
) : IdentityKeyCustodian, AutoCloseable {
    init {
        require(id.isNotBlank() && displayName.isNotBlank())
        require(kmsResourceUrl.protocol == URLProtocol.HTTPS && kmsResourceUrl.user == null && kmsResourceUrl.password == null)
        require(kmsResourceUrl.parameters.isEmpty() && kmsResourceUrl.fragment.isEmpty())
        require(kmsResourceUrl.encodedPath.trim('/').isNotEmpty() && !kmsResourceUrl.encodedPath.endsWith('/'))
    }

    private val client = httpClient.config { followRedirects = false; expectSuccess = false }

    /** Imports under the stable logical key ID and verifies the destination public key.
     * @param identity Original identity whose signing key is copied.
     * @param privateKey Original P-256 private JWK; never included in receipts or error messages. */
    override suspend fun importKey(identity: SigningIdentity, privateKey: EncodedKey.Jwk): IdentityCustodyReceipt {
        require(identity.keyId.matches(Regex("[A-Za-z0-9_-]{1,256}"))) { "Unsupported KMS key resource identifier" }
        val resource = "$kmsResourceUrl.${identity.keyId}"
        val privateBytes = privateKey.data.toByteArray()
        try {
            return client.preparePost("$resource/kms-service-api/keys/import/jwk") {
                setBody(TextContent(privateBytes.decodeToString(), ContentType.Application.Json))
            }.execute { response ->
                when (response.status.value) {
                    200, 201 -> Unit
                    401 -> throw IdentityProviderException(IdentityProviderFailure.InteractionRequired)
                    403 -> throw IdentityProviderException(IdentityProviderFailure.Rejected)
                    409 -> throw IdentityProviderException(IdentityProviderFailure.Conflict)
                    429, in 500..599 -> throw IdentityProviderException(IdentityProviderFailure.TemporarilyUnavailable)
                    else -> throw IdentityProviderException(IdentityProviderFailure.Rejected)
                }
                val bytes = response.bodyAsChannel().readRemaining(65_537).readByteArray()
                try {
                    if (bytes.size > 65_536) throw IdentityProviderException(IdentityProviderFailure.Rejected)
                    val view = Json.parseToJsonElement(bytes.decodeToString()).jsonObject
                    val jwk = view.getValue("key").jsonObject.getValue("jwk").jsonObject
                    // Enterprise local-JWK views can contain private material. Return only public members.
                    val publicJwk = buildJsonObject {
                        for (name in listOf("kty", "crv", "x", "y")) put(name, jwk.getValue(name))
                    }.toString()
                    val observed = EncodedKey.Jwk(BinaryData(publicJwk.encodeToByteArray()), false)
                    if (observed.toSpkiDer(KeySpec.Ec(EcCurve.P256)) != privateKey.toSpkiDer(KeySpec.Ec(EcCurve.P256)))
                        throw IdentityProviderException(IdentityProviderFailure.Conflict)
                    IdentityCustodyReceipt(resource, publicJwk)
                } catch (cause: IdentityProviderException) { throw cause }
                catch (_: Exception) { throw IdentityProviderException(IdentityProviderFailure.Rejected) }
                finally { bytes.fill(0) }
            }
        } catch (cause: CancellationException) { throw cause }
        catch (cause: IdentityProviderException) { throw cause }
        catch (_: Exception) { throw IdentityProviderException(IdentityProviderFailure.TemporarilyUnavailable) }
        finally { privateBytes.fill(0) }
    }

    /** Releases this adapter's configured HTTP client; the host retains ownership of its original client. */
    override fun close(): Unit = client.close()
}
